mtags/src/main/scala-2/scala/meta/internal/pc/InferredTypeProvider.scala

```diff
@@ -0,0 +1,185 @@
+import scala.meta.tokens.{Token => T}
+
+import org.eclipse.lsp4j.TextEdit
+
+/**
+ * Tries to calculate edits needed to insert the inferred type annotation
+ * in all the places that it is possible such as:
+ * - value or variable declaration
+ * - methods
+ * - pattern matches
+ * - for comprehensions
+ * - lambdas
+ *
+ * The provider will not check if the type does not exist, since there is no way to
+ * get that data from the presentation compiler. The actual check is being done via
+ * scalameta parser in InsertInferredType code action.
+ *
+ * @param compiler Metals presentation compiler
+ * @param params position and actual source
+ */
+final class InferredTypeProvider(
+    val compiler: MetalsGlobal,
+    params: OffsetParams
+) {
+  import compiler._
+
+  def inferredTypeEdits(): List[TextEdit] = {
+
+    val unit = addCompilationUnit(
+      code = params.text(),
+      filename = params.uri().toString(),
+      cursor = None
+    )
+
+    val pos = unit.position(params.offset)
+    val typedTree = typedTreeAt(pos)
+    val importPosition = autoImportPosition(pos, params.text())
+    val context = doLocateImportContext(pos, importPosition)
+    val history = new ShortenedNames(
+      lookupSymbol = name => context.lookupSymbol(name, _ => true) :: Nil
+    )
+
+    def additionalImports = importPosition match {
+      case None =>
+        // No import position means we can't insert an import without clashing with
+        // existing symbols in scope, so we just do nothing
+        Nil
+      case Some(importPosition) =>
+        history.autoImports(
+          pos,
+          context,
+          importPosition.offset,
+          importPosition.indent,
+          importPosition.padTop
+        )
+    }
+
+    def prettyType(tpe: Type) =
+      metalsToLongString(tpe.widen.finalResultType, history)
+
+    typedTree match {
+      /* `val a = 1` or `var b = 2`
+       *     turns into
+       * `val a: Int = 1` or `var b: Int = 2`
+       */
+      case vl @ ValDef(_, name, tpt, _) if !vl.symbol.isParameter =>
+        // it seems name might sometime contain the space afterwards
+        val nameEnd = tpt.pos.start + name.toString.trim().size
+        val nameEndPos = tpt.pos.withEnd(nameEnd).withStart(nameEnd).toLSP
+        val typeNameEdit = new TextEdit(nameEndPos, ": " + prettyType(tpt.tpe))
+        typeNameEdit :: additionalImports
+
+      /* `.map(a => a + a)`
+       *     turns into
+       * `.map((a: Int) => a + a)`
+       */
+      case vl @ ValDef(_, name, tpt, _) if vl.symbol.isParameter =>
+        val nameEnd = vl.pos.start + name.length()
+        val namePos = tpt.pos.withEnd(nameEnd).withStart(nameEnd).toLSP
+        def typeNameEdit(rightParen: String) =
+          new TextEdit(namePos, ": " + prettyType(tpt.tpe) + rightParen)
+
+        def paramStart = vl.pos.withEnd(vl.pos.start).toLSP
+        def startParamEdit = new TextEdit(paramStart, "(")
+
+        val needsParens = lastVisitedParentTrees match {
```

Comment:
So I think we'll need to maybe add another check here for this situation:

```scala
myList.map((a) => a * 2)
```
Because for whatever reason if someone has this, currently you are given:

```scala
myList.map(((a: Int)) => a * 2)
```

It's safe probably, but just an annoyance so it could be nice to catch it.

Reply:
Added a check and a test. Good catch!